cpp/include/cugraph/mtmg/resource_manager.hpp

```diff
@@ -0,0 +1,208 @@
+ * Based on how big a subset of the desired compute resources is desired, we can allocate some
+ * number of GPUs to the problem (up to the total set of managed resources).
+ *
+ * The returned instance can be used to create a graph, execute one or more algorithms, etc.  Once
+ * we are done the caller can delete the instance.
+ *
+ * At the moment, the caller is assumed to be responsible for scheduling use of the resources.
+ *
+ * For our first release, we will only consider a single node multi-GPU configuration, so the remote
+ * GPU methods are currently disabled via ifdef.
+ */
+class resource_manager_t {
+ public:
+  resource_manager_t() {}
+
+  /**
+   * @brief add a local GPU to the resource manager.
+   *
+   * @param rank       The rank to assign to the local GPU
+   * @param device_id  The device_id corresponding to this rank
+   */
+  void register_local_gpu(int rank, int device_id)
+  {
+    std::lock_guard<std::mutex> lock(lock_);
+
+    CUGRAPH_EXPECTS(local_rank_map_.find(rank) == local_rank_map_.end(),
+                    "cannot register same rank multiple times");
+
+    int num_gpus_this_node;
+    RAFT_CUDA_TRY(cudaGetDeviceCount(&num_gpus_this_node));
+
+    CUGRAPH_EXPECTS((device_id >= 0) && (device_id < num_gpus_this_node), "device id out of range");
+
+    local_rank_map_[rank] = device_id;
+  }
+
+#if 0
+  /**
+   * @brief add a remote GPU to the resource manager.
+   *
+   * FIXME: Need some sort of comms information for the remote GPU here...
+   *
+   * @param rank    The rank to assign to the remote GPU
+   */
+  void register_remote_gpu(int rank, TBD const& remote_info);
+#endif
+
+  /**
+   * @brief Create an instance using a subset of the registered resources
+   *
+   * The selected set of resources will be configured as an instance manager.
+   * If @ranks_to_include is a proper subset of the registered resources,
+   * ranks will be renumbered into the range [0, @p ranks_to_use.size()), making
+   * it a proper configuration.
+   *
+   * @param ranks_to_use        a vector containing the ranks to include in the instance.
+   *   Must be a subset of the entire set of available ranks.
+   * @param instance_manager_id a ncclUniqueId that is shared by all processes participating
+   *   in this instance.  All processes must use the same ID in this call, it is up
+   *   to the calling code to share this ID properly before the call.
+   *
+   * @return unique pointer to instance manager
+   */
+  std::unique_ptr<instance_manager_t> create_instance_manager(
+    std::vector<int> ranks_to_include, ncclUniqueId instance_manager_id) const
+  {
+    std::for_each(
+      ranks_to_include.begin(), ranks_to_include.end(), [local_ranks = local_rank_map_](int rank) {
+        CUGRAPH_EXPECTS(local_ranks.find(rank) != local_ranks.end(),
+                        "requesting inclusion of an invalid rank");
+      });
+
+    ncclGroupStart();
```

Comment:
I guess this is unnecessary if ncclCommIinitRank is called in different threads.

Reply:
Removed